### ISSUE TITLE
[dolphin]: view is not updated when dragndrop a folder from a zip from ark into home after its extraction without a refresh

### ISSUE THREAD

**Issue by star-buck** [blue-systems/plasma-5.9#71]

**Comment by notmart** [comment]
what does happen exactly? the folder that has been dragged from the zip to the home does not appear in dolphin?

**Comment by star-buck** [comment]
seems to work aka not reproduceable the second time, so will need to investigate more before knowing how to reliably get this behaviour...

**Comment by star-buck** [comment]
@notmart : yes, it only showed up after a F5 refresh (even after an initial 1 minute wait or so after the info message said "finished")

**Comment by star-buck** [comment]
i was able to record this on a freshly installed system:
https://www.dropbox.com/s/np6jruf9fgnfqxr/dolphin-refresh-2017-03-30_09-26-52.mkv?dl=0

**Comment by davidedmundson** [comment]
I've also managed to (accidentally) see it on my system the other way round.
I had a folder, I did right click compress -> here as zip  nothing happened, I refreshed and my new zip file was there.

I'll work on this solidly for today and tomorrow, but the lack of perfect reproducibility really doesn't make things easy.

I think my best option is to make a list of the expected code path from the kernel's signal to the model view; making a list of every method we should go down, and every possible branching point.
 
Then hopefully with a combo of debug and just code reviewing the crap out of everything we can find it...

**Comment by star-buck** [comment]
Sounds at least like a plan!

Note that I recall its almost always happening when files get changed,
removed or added for folders mapped via sambamounter, even though this way
of adding them should be transparent compared to smb://?

**Comment by davidedmundson** [comment]
So far I don't have anything concrete, but want to write up some things I've investigated, in case Marco's interested. 

The relevant code is far more complex than I would have imagined :/

Dolphin code is fine, the one time I managed to reproduce this with debug on, KCoreDirLister::itemsAdded wasn't called which means we need to look lower level.

I'll explain the relevant bits of code:
 - Dolphin gets directory listings and changes from KDirLister
 - KDirLister keeps a cache of 10 directories, though it will sometimes try to turn off monitoring for changes [1].
 - KDirLister gets updates from KDirWatch (See DirEntry)
 - KDirWatch can be temporarily inhibited by KDirWatch::stopDirWatch which happens whilst we're doing a copy/delete[2]
 - we follow the KDirWatch singleton so if any other chump comes along and calls KDirWatch::self()->removeDir(..) we'll stop the watcher even if we're using it.  [3]

KDirWatch::client -  a KDirWatch instance, watching one or more directories
KDirWAtch::entry - a directory being monitored by one or more clients

[1] We will have a bug if KDirLister gets it's item->incAutoUpdate/decAutoUpdate out of sync - then it will prematurely turn off tracking. 

We will also have a bug if we ever got two  KCoreDirListerCacheDirectoryData for the same directory without realising. This could happen if we manage to get two different URLs that both refer to the same absolute path? It's my current favourite theory.

[2] Though I've checked and every code path will always re-enable it, unless somehow the KJob just stalled without finishing. 
Also it doesn't force a rescan so potentially we could miss a file change if something happened whilst it was running, but the copyjob code paths don't seem relevant in this case. 

[3] I've checked everything that dolphin could possibly be loading. KSambaShare removes a watch, but only on Dolphin exit, so it's not that. There weren't many others.

**Comment by star-buck** [comment]
another instance showing dolphin not updating the view (really showing the two other issues besides the screenshot taken not appearing in this little snippet, that is the focus indicator and the left-click eaten while saving the screenshot):
https://www.dropbox.com/s/jxea09hs9bz5u0n/focus%20indicator%20-dolphin%20refresh-click%20eaten.mp4?dl=0

**Comment by davidedmundson** [comment]
Some good news.

This has been found and fixed by David Faure and Elvis Angellico and a somewhat lucky bug report.

Due to a completely different frameworks change we would sometimes end up watching "." In Linux terms that's the "current directory" of the folder the app was started from. By default with klauncher that ends up being /home/myusername

As I noted out above (but failed to see where this would happen) if we get two places watching the same directory but via different paths, it could miss updates.

Whether you see this bug or not depends on how you open dolphin! Something that was always different when I was looking for debug. 

Fix is in frameworks 5.35

**Comment by star-buck** [comment]
Awesome if really fixed now!
